File: baekjoon/java/1948.java
import java.util.*;

public class Main {
    static final int MAX_VERT = 10001;
    static int V, E, From, To;
    static List<Pair> adj[] = new ArrayList[MAX_VERT];
    static List<Pair> r_adj[] = new ArrayList[MAX_VERT];
    static int[] indegree, r_indegree;

    static class Pair {
        long weight;
        int to;

        Pair(long weight, int to) {
            this.weight = weight;
            this.to = to;
        }
    }

    // Function to show the vector (not used here but kept for debugging)
    static void showVec(int[] vec) {
        for (int i = 1; i <= V; i++) {
            System.out.print(vec[i] + " ");
        }
        System.out.println();
    }

    static void input() {
        Scanner sc = new Scanner(System.in);
        V = sc.nextInt();
        E = sc.nextInt();
        indegree = new int[V + 1];
        r_indegree = new int[V + 1];
        
        for (int i = 1; i <= V; i++) {
            adj[i] = new ArrayList<>();
            r_adj[i] = new ArrayList<>();
        }
        
        for (int i = 0; i < E; i++) {
            int from = sc.nextInt();
            int to = sc.nextInt();
            long weight = sc.nextLong();
            indegree[to]++;
            r_indegree[from]++;
            adj[from].add(new Pair(weight, to));
            r_adj[to].add(new Pair(weight, from));
        }
        From = sc.nextInt();
        To = sc.nextInt();
    }

    static long[] getDistance(List<Pair>[] adj, int[] indegree, int start) {
        long[] res = new long[V + 1];
        Arrays.fill(res, 0);
        Queue<Integer> q = new LinkedList<>();
        q.add(start);

        while (!q.isEmpty()) {
            int node = q.poll();

            for (Pair ele : adj[node]) {
                long cost = ele.weight;
                int to = ele.to;

                res[to] = Math.max(res[to], res[node] + cost);
                indegree[to]--;
                if (indegree[to] == 0)
                    q.add(to);
            }
        }
        return res;
    }

    static int getResult(long[] distance, long[] r_distance) {
        int res = 0;
        for (int i = 1; i <= V; i++) {
            for (Pair ele : adj[i]) {
                int j = ele.to;
                long v = ele.weight;

                if (distance[i] + r_distance[j] + v == distance[To]) {
                    res++;
                }
            }
        }
        return res;
    }

    static void solve() {
        input();
        long[] Distance = getDistance(adj, indegree, From);
        long[] R_Distance = getDistance(r_adj, r_indegree, To);
        int edgeNum = getResult(Distance, R_Distance);

        System.out.println(Distance[To]);
        System.out.println(edgeNum);
    }

    public static void main(String[] args) {
        solve();
    }
}

File: baekjoon/C++/1948.cpp
#include <bits/stdc++.h>
#define MAX_VERT 10001
#define endl '\n'
using namespace std;
using ull = unsigned long long;
using pui = pair<unsigned long, int>; // weight to

int V, E, From, To;
vector<pui> adj[MAX_VERT], r_adj[MAX_VERT];
vector<int> indegree, r_indegree;


void show_vec(const auto& vec) noexcept
{
	for (int i = 1; i <= V; i++)
	{
		cout << vec[i] << " ";
	}cout << endl;
}

void input() noexcept
{
	int from, to;
	ull weight;
	cin >> V >> E;
	indegree.assign(V+1, 0); r_indegree.assign(V+1, 0);
	for (int i = 0; i < E; i++)
	{
		cin >> from >> to >> weight;
		++indegree[to];
		++r_indegree[from];
		adj[from].push_back(make_pair(weight, to));
		r_adj[to].push_back(make_pair(weight, from));
	}
	cin >> From >> To;
}

vector<ull> get_distance(vector<pui> adj[],vector<int>& indegree,  int& start) noexcept
{
	
	vector<ull> res(V+1,0);
	queue<int> q;
	q.push(start);

	while (!q.empty())
	{
		int node = q.front();
		q.pop();

		for(auto& ele: adj[node])
		{	
			ull cost = ele.first;
			int to = ele.second;
	
			res[to] = max(res[to], res[node] + cost);
			indegree[to]--;
			if (indegree[to] == 0)
				q.push(to);
		}
	}
	return res;
}

int get_result(vector<ull>& distance, vector<ull>& r_distance) noexcept
{
	int res = 0;
	for (int i = 1; i <= V; i++)
	{
		for(auto&ele: adj[i])
		{	
			int j = ele.second;
			ull v = ele.first;

			if (distance[i] + r_distance[j] + v == distance[To])
				++res;
		}
	}
	return res;
}

void solve() noexcept
{
	input();
	// show_vec(indegree);
	// show_vec(r_indegree);
	auto Distance = get_distance(adj,indegree ,From);
	auto R_Distance = get_distance(r_adj,r_indegree, To);
	// show_vec(Distance);
	// show_vec(R_Distance);
	auto edge_num = get_result(Distance, R_Distance);

	cout << Distance[To] << endl << edge_num << endl;
}


int main(void)
{
	ios::sync_with_stdio(0);
	cin.tie(0); cout.tie(0);

	solve();

	return (0);
}